File: [2] Calculations/Self-Checkout/SelfCheckout.py
# Create a simple self-checkout system.
# Prompt for the prices and quantities of three items.
# Calculate the subtotal of the items.
# Then calculate the tax using a tax rate of 5.5%.
# Print out the line items with the quantity and total,
# and then print out the subtotal, tax amount, and total.


# Revise the program to ensure that prices and quantities are entered as numeric values.
# Don’t allow the user to proceed if the value entered is not numeric.
def ValidatedNumber(prompt, _min=None, _max=None, _type=None):
    while True:
        user_input = input(prompt)
        # Be sure you explicitly convert input to numerical data before doing any calculations.
        try:
            number = float(user_input)
            if (_min is None or _min <= number) and (_max is None or _max >= number):
                return _type == "float" and number or int(number)
            else:
                print(f"Input out of range. [{_min}-{_max}]")
        except ValueError:
            print("! Error: Input is NaN")


TAX = 0.055


# Keep the input, processing, and output parts of your program separate.
# Collect the input, then do the math operations and string building,
# and then print out the output.

class Item:
    def __init__(self, _i):
        self.price = ValidatedNumber(f"Enter the price of item {_i}: ")
        self.quantity = ValidatedNumber(f"Enter the quantity of item {_i}: ")
        self.subtotal = self.price * self.quantity


class Basket:
    def __init__(self, _items):
        self.items = _items
        self.subtotal = 0.0
        for each in self.items:
            self.subtotal += each.subtotal
        self.tax = self.subtotal * TAX
        self.total = self.subtotal + self.tax

    def display_totals(self):
        print(f"Subtotal: ${round(self.subtotal, 2)}")
        print(f"Tax: ${round(self.tax, 2)}")
        print(f"Total: ${round(self.total, 2)}")


items = []
i = 1
item_count = ValidatedNumber("How many items do you have? ", _type="int")
while i <= item_count:
    items.append(Item(i))
    i += 1
checkout = Basket(items)
checkout.display_totals()

File: [2] Calculations/Self-Checkout/SelfCheckout.cs
using System;
using System.Collections.Generic;
using static CSharp._2__Calculations.DataValidation;

namespace CSharp._2__Calculations.Self_Checkout
{
    internal class SelfCheckout
    {
        private const float TAX = 0.055f;

        private class Item
        {
            private readonly float? price, quantity;
            public float? subtotal;
            public Item(int _i)
            {
                price = ValidatedNumber($"Enter the price of item {_i}: ", _sentinel: "STOP");
                if (price == null)
                    throw new Exception();
                
                quantity = ValidatedNumber($"Enter the quantity of item {_i}: ", _sentinel: "STOP");
                if (quantity == null)
                    throw new Exception();
                subtotal = price * quantity;
            }
        }

        private class Basket
        {
            private readonly List<Item> items;
            private readonly float subtotal, tax, total;
            public Basket(List<Item> _items)
            {
                items = _items;
                foreach (Item item in items)
                    subtotal += (float)item.subtotal;
                tax = subtotal * TAX; 
                total = subtotal + tax;
            }

            public void DisplayTotals()
            {
                Console.WriteLine($"Subtotal: ${subtotal:F2}");
                Console.WriteLine($"Tax: ${tax:F2}");
                Console.WriteLine($"Total: ${total:F2}");
            }
        }

        static void Main8()
        {
            List<Item> items = [];
            for (int i = 1; true; i++)
            {
                try { items.Add(new Item(i)); }
                catch { break; }
            }
            Basket checkout = new(items);
            checkout.DisplayTotals();
        }
    }
}
